rfcs/side-effect-lifecycle-scheduling.md

```diff
@@ -0,0 +1,124 @@
+
+>Draft
+
+# RFC: Encoding Side Effect Lifecycles as Subgraphs
+
+This RFC presents an API for scheduling around the lifecycle of side effects. This assumes that exclusive systems can be labeled and ordered in the same way parallel systems can and does not include how the executor might need to be changed to support this API.
+
+## Motivation
+
+A system may produce a side effect that other systems must read, write, and potentially consume. Systems that do this have a very specific order since the side effect must be produced before it can be acted on and cannot be acted upon after they have been consumed. An example of this would be commands. Commands are created in one system and then consumed later when they're applied to the world. If you tried to read the command queue after the commands have been applied, you would only find an empty queue.
+
+We can express this required ordering with labels, before, and after, but they lack semantic meaning. This is an api for creating subgraphs that express the required ordering between producers, consumers, readers, and writers.
+
+## User Facing Description
+
+### Scheduling Side Effects API
+* `.produces(Label)` System produces a side effect of type `Label`
+* `.consumes(Label)` No read/modifies systems can run after this system. An example that would use this is the system that applies commands to the world.
+* `.reads(Label)` system has read access to side effect and must run after the producers and before the consumer. An example of this type of system would be systems that depend on a run criteria.
+* `.writes(Label)` system writes to the side effect and must run after the producers and before the consumer
+
+#### Relationships enforced between producers, readers, modifiers, and consumers
+
+* Producers must be before readers, writers, and the consumer.
+* Readers and writers must be after the producer and before the consumer. They can run in any order in relation to other readers and writers. If there needs to be an order between readers and writers, use the normal before and after labeling.
+* Consumers are run after producers, readers, and writers. There should only be one consumer of a type of side effect.
+
+### Some Simple Examples
+
+#### Scheduling around command application
+```rs
+fn main() {
+    App::new()
+        .add_system(apply_commands_buffer.consumes("commands"))
+        // if we don't care about when the commands are applied
+        // we can ignore the new api
+        .add_system(create_commands_A)
+        
+        // if we need to schedule a system after commands are applied
+        .add_system(create_commands_A.label("system A").produces("commands"))
+        .add_system(after_commands_A.after("system A").after("commands"))
```

Comment:
I'm not sure this will give the desired behavior: it will run after *all* of the available systems with the "commands" label are complete, when we want "at least one".

Reply:
This can be particularly nasty if you want to chain several of these things back-to-back within a frame.